File: go/storage/jsonl_log.go
// Package storage provides persistence backends for the agent memory system.
package storage

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	agentmemory "github.com/CodeHalwell/Memories/go"
)

// JSONLLogger is an append-only JSONL logger for raw agent outputs.
// This is the immutable ground truth — entries are never modified or deleted.
type JSONLLogger struct {
	LogDir string
}

// NewJSONLLogger creates a JSONLLogger at the given directory.
func NewJSONLLogger(logDir string) (*JSONLLogger, error) {
	if err := os.MkdirAll(logDir, 0o755); err != nil {
		return nil, fmt.Errorf("creating log dir: %w", err)
	}
	return &JSONLLogger{LogDir: logDir}, nil
}

func (l *JSONLLogger) sessionPath(sessionID string) string {
	safeID := filepath.Base(sessionID)
	return filepath.Join(l.LogDir, safeID+".jsonl")
}

// Append writes an entry and returns (filePath, byteOffset).
func (l *JSONLLogger) Append(entry agentmemory.RawLogEntry) (string, int64, error) {
	path := l.sessionPath(entry.SessionID)

	data, err := json.Marshal(entry)
	if err != nil {
		return "", 0, fmt.Errorf("marshalling entry: %w", err)
	}
	line := append(data, '\n')

	var byteOffset int64
	if info, err := os.Stat(path); err == nil {
		byteOffset = info.Size()
	}

	f, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return "", 0, fmt.Errorf("opening log file: %w", err)
	}
	defer f.Close()

	if _, err := f.Write(line); err != nil {
		return "", 0, fmt.Errorf("writing entry: %w", err)
	}

	return path, byteOffset, nil
}

// ReadEntry reads a single entry at the given byte offset.
func (l *JSONLLogger) ReadEntry(filePath string, byteOffset int64) (agentmemory.RawLogEntry, error) {
	var entry agentmemory.RawLogEntry

	f, err := os.Open(filePath)
	if err != nil {
		return entry, fmt.Errorf("opening log file: %w", err)
	}
	defer f.Close()

	if _, err := f.Seek(byteOffset, 0); err != nil {
		return entry, fmt.Errorf("seeking to offset: %w", err)
	}

	buf := make([]byte, 0, 4096)
	tmp := make([]byte, 1)
	for {
		n, err := f.Read(tmp)
		if n > 0 {
			if tmp[0] == '\n' {
				break
			}
			buf = append(buf, tmp[0])
		}
		if err != nil {
			break
		}
	}

	if err := json.Unmarshal(buf, &entry); err != nil {
		return entry, fmt.Errorf("unmarshalling entry: %w", err)
	}
	return entry, nil
}

// IterSession yields all entries for a session in order via a callback.
func (l *JSONLLogger) IterSession(sessionID string, fn func(agentmemory.RawLogEntry) bool) error {
	path := l.sessionPath(sessionID)
	file, err := os.Open(path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return fmt.Errorf("opening session file: %w", err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		var entry agentmemory.RawLogEntry
		if err := json.Unmarshal([]byte(line), &entry); err != nil {
			continue
		}
		if !fn(entry) {
			break
		}
	}
	if err := scanner.Err(); err != nil {
		return fmt.Errorf("scanning session file: %w", err)
	}
	return nil
}

// Search performs a case-insensitive text search within a session log.
func (l *JSONLLogger) Search(sessionID, text string) ([]agentmemory.RawLogEntry, error) {
	var results []agentmemory.RawLogEntry
	lower := strings.ToLower(text)

	err := l.IterSession(sessionID, func(entry agentmemory.RawLogEntry) bool {
		if strings.Contains(strings.ToLower(entry.Content), lower) {
			results = append(results, entry)
		}
		return true
	})
	return results, err
}

// ListSessions returns all session IDs that have log files.
func (l *JSONLLogger) ListSessions() ([]string, error) {
	entries, err := filepath.Glob(filepath.Join(l.LogDir, "*.jsonl"))
	if err != nil {
		return nil, err
	}
	var sessions []string
	for _, e := range entries {
		base := filepath.Base(e)
		sessions = append(sessions, strings.TrimSuffix(base, ".jsonl"))
	}
	return sessions, nil
}

// SessionSize returns the file size in bytes for a session log.
func (l *JSONLLogger) SessionSize(sessionID string) int64 {
	path := l.sessionPath(sessionID)
	info, err := os.Stat(path)
	if err != nil {
		return 0
	}
	return info.Size()
}

File: rust/src/storage/jsonl_log.rs
//! Raw JSONL log writer and indexer.
//!
//! Every agent output is appended to a session-specific JSONL file. This is the
//! immutable ground truth — entries are never modified or deleted.

use std::path::{Path, PathBuf};

use crate::config::default_log_dir;
use crate::models::RawLogEntry;

/// Append-only JSONL logger for raw agent outputs.
pub struct JSONLLogger {
    log_dir: PathBuf,
}

impl JSONLLogger {
    pub fn new(log_dir: Option<PathBuf>) -> Result<Self, std::io::Error> {
        let dir = log_dir.unwrap_or_else(default_log_dir);
        std::fs::create_dir_all(&dir)?;
        Ok(Self { log_dir: dir })
    }

    fn session_path(&self, session_id: &str) -> PathBuf {
        // Sanitize session_id to prevent path traversal
        let safe_id = Path::new(session_id)
            .file_name()
            .map(|n| n.to_string_lossy().into_owned())
            .unwrap_or_else(|| session_id.to_string());
        self.log_dir.join(format!("{safe_id}.jsonl"))
    }

    /// Append an entry and return (file_path, byte_offset).
    pub fn append(&self, entry: &RawLogEntry) -> Result<(String, u64), std::io::Error> {
        let path = self.session_path(&entry.session_id);
        let line = serde_json::to_string(entry).map_err(|e| {
            std::io::Error::new(std::io::ErrorKind::InvalidData, e)
        })?;

        let byte_offset = if path.exists() {
            std::fs::metadata(&path)?.len()
        } else {
            0
        };

        use std::io::Write;
        let mut file = std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(&path)?;
        writeln!(file, "{line}")?;

        Ok((path.to_string_lossy().into_owned(), byte_offset))
    }

    /// Read a single entry at the given byte offset.
    pub fn read_entry(&self, file_path: &str, byte_offset: u64) -> Result<RawLogEntry, Box<dyn std::error::Error>> {
        use std::io::{BufRead, Seek, SeekFrom};
        let mut file = std::fs::File::open(file_path)?;
        file.seek(SeekFrom::Start(byte_offset))?;
        let mut reader = std::io::BufReader::new(file);
        let mut line = String::new();
        reader.read_line(&mut line)?;
        let entry: RawLogEntry = serde_json::from_str(line.trim())?;
        Ok(entry)
    }

    /// Yield all entries for a session in order.
    pub fn iter_session(&self, session_id: &str) -> Result<Vec<RawLogEntry>, Box<dyn std::error::Error>> {
        let path = self.session_path(session_id);
        if !path.exists() {
            return Ok(Vec::new());
        }
        use std::io::BufRead;
        let file = std::fs::File::open(path)?;
        let reader = std::io::BufReader::new(file);
        let mut entries = Vec::new();
        for line in reader.lines() {
            let line = line?;
            let trimmed = line.trim();
            if !trimmed.is_empty() {
                let entry: RawLogEntry = serde_json::from_str(trimmed)?;
                entries.push(entry);
            }
        }
        Ok(entries)
    }

    /// Simple text search within a session log.
    pub fn search(&self, session_id: &str, text: &str) -> Result<Vec<RawLogEntry>, Box<dyn std::error::Error>> {
        let entries = self.iter_session(session_id)?;
        let lower_text = text.to_lowercase();
        let results = entries
            .into_iter()
            .filter(|e| e.content.to_lowercase().contains(&lower_text))
            .collect();
        Ok(results)
    }

    /// Return all session IDs that have log files.
    pub fn list_sessions(&self) -> Result<Vec<String>, std::io::Error> {
        let mut sessions = Vec::new();
        if self.log_dir.exists() {
            let mut entries: Vec<_> = std::fs::read_dir(&self.log_dir)?
                .filter_map(|e| e.ok())
                .filter(|e| {
                    e.path()
                        .extension()
                        .is_some_and(|ext| ext == "jsonl")
                })
                .collect();
            entries.sort_by_key(|e| e.path());
            for entry in entries {
                if let Some(stem) = entry.path().file_stem() {
                    sessions.push(stem.to_string_lossy().into_owned());
                }
            }
        }
        Ok(sessions)
    }

    /// Return the file size in bytes for a session log.
    pub fn session_size(&self, session_id: &str) -> u64 {
        let path = self.session_path(session_id);
        if path.exists() {
            std::fs::metadata(&path).map(|m| m.len()).unwrap_or(0)
        } else {
            0
        }
    }
}
